dsl/template.go

```diff
@@ -0,0 +1,108 @@
+package dsl
+
+import (
+	"fmt"
+
+	"github.com/flosch/pongo2/v6"
+	"github.com/goccy/go-json"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/function"
+)
+
+// TemplateFunc is a stateful cty function that evaluates a file and variables
+// using the pongo2 library, which matches the Django template language. It
+// returns the results as a string. It finds the file relative to the current
+// .hops file. The data of the file is determined at startup time.
+//
+// It is stateful because it requires the HopsFiles struct and the hopsDirectory
+// to be passed in.
+func TemplateFunc(hops *HopsFiles, hopsDirectory string) function.Function {
+	return function.New(&function.Spec{
+		Params: []function.Parameter{
+			{
+				Name: "filename",
+				Type: cty.String,
+			},
+			{
+				Name: "variables",
+				Type: cty.DynamicPseudoType,
+			},
+		},
+		Type: function.StaticReturnType(cty.String),
+		// Closure over hops and hopsDirectory
+		Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+			filenameVal := args[0]
+			filename := filenameVal.AsString()
+			variablesVal := args[1]
+
+			var variables map[string]interface{}
+
+			// Check if variablesVal is a string and try to json.Unmarshal it
+			if variablesVal.Type() == cty.String {
```

Comment:
I'm not keen on this, it's an extra complication that adds a gotcha.

I'd just take the argument as a typed object and drop this implicit string to json conversion. They can already convert it to an object via `jsondecode` if that's what they meant, but the logic above means they have no choice in the matter. Would be frustrating for users if they ever actually want to pass a json string through.

Reply:
Yeah, no problem. Makes sense. I would say, they can't pass just a json string, because the template can't handle that.

Reply:
Removed.